### ISSUE TITLE
WebAPI dont support mutliple return type.

### ISSUE THREAD

**Issue by kuldeepdhaka** [magento/magento2#18242]
### Preconditions

1. 2.3-develop

### Steps to reproduce

```
interface ErrorFetchResponceInterface {
	/**
	 * @return string
	 */
	public function getMessage();
}

interface SuccessFetchResponceInterface {
	/**
	 * @return string
	 */
	public function getLicence();

	/**
	 * @return int
	 */
	public function getRemaining();
}

interface MyRestApiInterface1 {
	/**
	 * @param string $license
	 * @return \MadResistor\MyModule\Api\ErrorFetchResponceInterface
	 * @return \MadResistor\MyModule\Api\SuccessFetchResponceInterface
	 */
	public function fetch($license);
}

interface MyRestApiInterface1Variant {
	/**
	 * @param string $license
	 * @return \MadResistor\MyModule\Api\ErrorFetchResponceInterface|\MadResistor\MyModule\Api\SuccessFetchResponceInterface
	 */
	public function fetch($license);
}


interface MyRestApiInterface2 {
	/**
	 * @param string $license
	 * @return mixed
	 */
	public function fetch($license);
}
```

`MyRestApiInterface1`:
If `fetch()` return `ErrorFetchResponceInterface` everything is fine.
BUT, 
if `fetch()` return type of `SuccessFetchResponceInterface`, then Magento try to search for method `getMessage()` that exists only in `ErrorFetchResponceInterface`.

`MyRestApiInterface1Variant` Dont work
`MyRestApiInterface2`: `Class mixed does not exist`

### Expected result

Based on class of object returned by `fetch()` (that implements one of the interfaces mentioned in `@return`), perform the same conversion but for that interface.

### Actual result

Basically Magento is using the first `@return` type for reference and not referencing other than first.

https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Webapi/ServiceOutputProcessor.php#L56-L60
https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Webapi/ServiceOutputProcessor.php#L92-L117

**Comment by magento-engcom-team** [comment]
Hi @kuldeepdhaka. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@kuldeepdhaka do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by kuldeepdhaka** [comment]
`
return current($methodDocBlock->getTags('return'));
`
This is the culprit block that throw away all other return types:
https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Reflection/TypeProcessor.php#L747

**Comment by kuldeepdhaka** [comment]
Temporary workaround
==================

```
interface FetchResponceInterface {
	/**
	 * @return null
	 */
	public function getMessage();

	/**
	 * @return null
	*/
	public function getLicence();

	/**
	 * @return null
	*/
	public function getRemaining();
}

class FetchResponce implements FetchResponceInterface {
	protected $_data;

	public function __construct($data) {
		$this->_data = $data;
	}

	protected function getData($key) {
		if (array_key_exists($key, $this->_data)) {
			return $this->_data[$key];
		}
	}

	public function getMessage() { $this->getData['message']; }
	public function getLicence()  { $this->getData['licence']; }
	public function getRemaining()  { $this->getData['remaining']; }
}

interface MyRestApiInterface {
	/**
	 * @param string $license
	 * @return \MadResistor\MyModule\Api\FetchResponceInterface
	 */
	public function fetch($license);
}

```

After exploring the code more, if a function define its `@return` type as `null` and only return if data is available, it will simply build a json object with returned types.
Reference: https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Reflection/DataObjectProcessor.php#L82-L89

Note: **Only tested for scalars (int,string...)**

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @kuldeepdhaka thank you for you report, It will be processed faster if you move this to https://github.com/magento/community-features

Best regards!
